Add examples for kubectl describe, make gendocs less spammy

j3ffml opened the pull request:
Added examples to `kubectl describe` and modified  `hack/{run/verify}-gendocs.sh` to ignore generated-timestamp-only changes so minor edits to kubectl code don't have to touch the entire set of md docs.

jlowdermilk commented:
cc @eparis, @filbranden for bash

filbranden commented on hack/lib/util.sh:
```diff
@@ -139,6 +135,23 @@ kube::util::gen-doc() {
   ls "${tmpdir}" | LC_ALL=C sort > "${tmpdir}/.files_generated"
+
+  # remove all old generated file from the destination
+  for file in $(cat "${dest}/.files_generated" 2>/dev/null); do
```
This is usually bad form (`for i in $(cat ...)`), though I can see it is pre-existing here.

If what you have is one file name per line, then consider the superior:

```shell
  while read file; do
    ...
  done <"${dest}/.files_generated"
```

jlowdermilk replied:
Done.

filbranden commented on hack/lib/util.sh:
```diff
@@ -139,6 +135,23 @@ kube::util::gen-doc() {
+
+  # remove all old generated file from the destination
+  for file in $(cat "${dest}/.files_generated" 2>/dev/null); do
+    if [[ -e "${tmpdir}/${file}" && ! -z "${skipline}" ]]; then
```
Use `-n` instead of `! -z` to test that it is not empty.

jlowdermilk replied:
Done.

filbranden commented on hack/lib/util.sh:
```diff
@@ -139,6 +135,23 @@ kube::util::gen-doc() {
+  # remove all old generated file from the destination
+  for file in $(cat "${dest}/.files_generated" 2>/dev/null); do
+    if [[ -e "${tmpdir}/${file}" && ! -z "${skipline}" ]]; then
+      # If RCS format diff has 3 lines, only 1 line changed.
```
Not necessarily, if you added two lines without removing any then the diff would also have 3 lines. What is it that you are trying to check here?

jlowdermilk replied:
Ah bother. This is trying to check that only one line was changed, and that said line matches the general grep regexp `${skipline}`. I'll see if I can use `--LTYPE-line-format` to achieve that.

filbranden commented on hack/lib/util.sh:
```diff
@@ -139,6 +135,23 @@ kube::util::gen-doc() {
+    if [[ -e "${tmpdir}/${file}" && ! -z "${skipline}" ]]; then
+      # If RCS format diff has 3 lines, only 1 line changed.
+      local fdiff=$(diff -n "${dest}/${file}" "${tmpdir}/${file}")
+      if [[ $(echo "${fdiff}" | wc -l) == 3 && $(echo "${fdiff}" | grep -e "${skipline}") ]]; then
```
What is in `${skipline}`? Is it really supposed to be a general grep regexp? Or is it a fixed string? Is it supposed to match a whole line?

This command is wrong, since the second part will actually evaluate the result of `$(echo "${fdiff}" | grep -e "${skipline}")` and that does not make much sense, I think what you wanted was to see whether the `grep` matched or not.

I'm quite confused by what this is trying to do...

jlowdermilk replied:
`${skipline}` is just a general grep regexp. The purpose of this is to detect files where the generated output does no meaningful change, it just updates the timestamp
```
-###### Auto generated by spf13/cobra at 2015-04-17 18:59:11.486604006 +0000 UTC
+###### Auto generated by spf13/cobra at 2015-04-23 20:38:18.472875747 +0000 UTC
```

filbranden replied:
In which case, may I suggest filtering out lines starting with `"^###### Auto generated by spf13/cobra at"` and then comparing that the contents are exactly the same? Or instead of filtering them out, replacing them with something like `"###### Auto generated by spf13/cobra at XXXX-XX-XX XX:XX:XX.XXXXXXXXX +XXXX XXX"`.

If your use case is fairly specific, I think that's a lot superior than trying to interpret diff output from bash code... Maybe replace "skipline" with "prefix" and assume you know what the regexp for a timestamp you recognize is or should be?

jlowdermilk replied:
Changed skipline to prefix, and filtering file contents by `${prefix}` before comparing that contents are the same. PTAL

jlowdermilk replied:
I assumed diff was more efficient, not that it matters. Done.

filbranden commented on hack/lib/util.sh:
```diff
@@ -139,6 +135,23 @@ kube::util::gen-doc() {
+        mv "${dest}/${file}" "${tmpdir}/${file}"
+      fi
+    else
+      set +e
```
No, don't disable it, even temporarily...

Just use a `||` on the line you're OK with failing, for example:

```shell
  rm "${dest}/${file}" || :
```

Or the slightly more verbose:

```shell
  rm "${dest}/${file}" || true
```

filbranden commented on hack/lib/util.sh:
```diff
@@ -139,6 +135,21 @@ kube::util::gen-doc() {
+  # remove all old generated file from the destination
+  while read file; do
+    if [[ -e "${tmpdir}/${file}" && -n "${skipprefix}" ]]; then
+      local original=$(cat "${dest}/${file}" | grep -v "${skipprefix}")
```
Minor problem here with `local` discarding the error result of `$(...)`, so ideally what you want is:

```shell
  local original generated
  original=$(cat "${dest}/${file}" | grep -v "${skipprefix)")
```

Though you actually want that to work even if `grep` fails (which happens if the initial file is empty or made of only lines matching ${skipprefix}!) so you want the `|| :` there. You can also combine the `cat` and `grep` since the `cat` isn't really doing anything there... You could also assume that the prefix is a prefix (though it's still a regexp) and actually add the `^` here.

You also are doing both `original` and `generated` from ${dest}, I think one of them was supposed to be ${tmpdir}.

```shell
  local original generated
  original=$(grep -v "^${skipprefix}" "${tmpdir}/${file}") || :
  generated=$(grep -v "^${skipprefix}" "${dest}/${file}") || :
```

filbranden commented on hack/lib/util.sh:
```diff
@@ -139,6 +135,21 @@ kube::util::gen-doc() {
+    if [[ -e "${tmpdir}/${file}" && -n "${skipprefix}" ]]; then
+      local original=$(cat "${dest}/${file}" | grep -v "${skipprefix}")
+      local generated=$(cat "${dest}/${file}" | grep -v "${skipprefix}")
+      if [[ -z $(diff <(echo "${original}") <(echo "${generated}")) ]]; then
```
You already have the two strings in memory and you don't really use the diff restuls... why not just the below?

```shell
  if [[ "${original}" == "${generated"} ]]l; then
    ...
```

j3ffml committed:
Add examples for kubectl describe, and make gendocs less spammy

Makes hack/{run/verify}-gendocs.sh ignore generated-timestamp-only
changes so minor changes to kubectl don't have to touch the entire
set of md docs.

filbranden commented:
Looks good for bash! Thanks for addressing the items in the review! :-)

Let me know if you'd like me to merge it or if you have other reviewers in mind or are still running some extra tests to ensure this actually does what it states.

Cheers,
Filipe

filbranden commented on hack/lib/util.sh:
```diff
@@ -119,17 +119,13 @@ kube::util::wait-for-jobs() {
-# takes a binary to run $1 and then copies the results to $2
+# Takes a binary to run $1 and then copies the results to $2.
+# If the generated and original files are the same after filtering lines
+# that match $3, copy is skipped.
```
Just one thing is I think you could improve documentation here to indicate what $3 is exactly. Some examples might help. Though now it's fairly clear from the code, so I think someone reading the code will understand what the intent is here.

jlowdermilk commented:
Thanks for helping fix my dreadful bash. This should be good to merge now, but also fine to label lgtm and merge later.

filbranden commented:
Always happy to help!

LGTM, feel free to self merge or have some other eyes take a look on it if you'd like.

Cheers,
Filipe
